### PR TITLE
Add FlyingThings3D dataset for optical flow

### PR DESCRIPTION
**Pull request by NicolasHug**
This PRs adds the FlyingThings3D dataset for optical flow

### CODE REVIEW

File: torchvision/datasets/_optical_flow.py [hunk 1]
```diff
@@ -208,6 +210,89 @@ def _read_flow(self, file_name):
         return _read_16bits_png_with_flow_and_valid_mask(file_name)
 
 
+class FlyingThings3D(FlowDataset):
+    """`FlyingThings3D <https://lmb.informatik.uni-freiburg.de/resources/datasets/SceneFlowDatasets.en.html>`_ dataset for optical flow.
+
+    The dataset is expected to have the following structure: ::
+
+        root
+            FlyingThings3D
```
Comment (NicolasHug) [inline torchvision/datasets/_optical_flow.py:219]: TODO

File: test/test_datasets.py [hunk 2]
```diff
@@ -1998,5 +1998,78 @@ def test_bad_input(self):
...
+        splits = ("TRAIN", "TEST")
+        letters = ("A", "B", "C")
+        subfolders = ("0000", "0001")
+        cameras = ("left", "right")
+        for pass_name in passes:
+            for split in splits:
+                for letter in letters:
+                    for subfolder in subfolders:
+                        current_folder = root / pass_name / split / letter / subfolder
+                        for camera in cameras:
```
Comment (pmeier) [inline test/test_datasets.py:2024]: That looks kind of horrifying. Maybe `itertools.product`?

File: test/test_datasets.py [hunk 3]
```diff
@@ -1998,5 +1998,78 @@ def test_bad_input(self):
...
+            for letter in letters:
+                for subfolder in subfolders:
+                    for direction in directions:
+                        current_folder = root / "optical_flow" / split / letter / subfolder / direction
+                        for camera in cameras:
+                            os.makedirs(str(current_folder / camera))
+                            for i in range(num_images_per_camera):
+                                datasets_utils.make_fake_pfm_file(
+                                    self.FLOW_H, self.FLOW_W, file_name=str(current_folder / camera / f"{i}.pfm")
+                                )
```
Comment (pmeier) [inline test/test_datasets.py:2043]: Same here.

File: test/test_datasets.py [hunk 4]
```diff
@@ -1998,5 +1998,78 @@ def test_bad_input(self):
...
+        with self.create_dataset(config=config) as (dataset, _):
+            assert dataset._flow_list and len(dataset._flow_list) == len(dataset._image_list)
+            for _, _, flow in dataset:
+                assert flow.shape == (2, self.FLOW_H, self.FLOW_W)
+                # We don't check the values because the reshaping and flipping makes it hard to figure out
+
+    def test_bad_input(self):
+        with pytest.raises(ValueError, match="split must be either"):
+            with self.create_dataset(split="bad"):
+                pass
```
Comment (pmeier) [inline test/test_datasets.py:2063]: If you go with my suggestion from https://github.com/pytorch/vision/pull/4860#discussion_r743441116, we probably don't need these tests.

File: torchvision/datasets/_optical_flow.py [hunk 5]
```diff
@@ -208,6 +210,89 @@ def _read_flow(self, file_name):
...
+
+    Args:
+        root (string): Root directory of the Sintel Dataset.
```
Comment (pmeier) [inline torchvision/datasets/_optical_flow.py:222]:
```suggestion
        root (string): Root directory of the FlyingThings3D Dataset.
```

File: torchvision/datasets/_optical_flow.py [hunk 6]
```diff
@@ -208,6 +210,89 @@ def _read_flow(self, file_name):
...
+
+        if camera not in ("left", "right", "both"):
+            raise ValueError("camera must be either 'left', 'right', or 'both'")
+        cameras = ["left", "right"] if camera == "both" else [camera]
+
+        root = Path(root) / "FlyingThings3D"
+
+        for pass_name in passes:
+            for camera in cameras:
+                for direction in ["into_future", "into_past"]:
```
Comment (pmeier) [inline torchvision/datasets/_optical_flow.py:256]: Same as above.

**Review by pmeier — APPROVED** [review]
Some minor comments inline plus you can apply the same comments I had in #4860 here. Otherwise LGTM, thanks @NicolasHug!

File: test/test_datasets.py (hunk 4 above)
Comment (NicolasHug) [reply test/test_datasets.py]: I'll go with your suggestion but still keep the test 🤓

**Pull request merged**